### ISSUE TITLE
Automate building the documentation

### ISSUE THREAD

**Issue by berquist** [cclib/cclib.github.io#32]
* Documentation configuration scripts, rst files, etc. should be moved to the code repo so that both can be updated simultaneously, which is necessary when adding/modifying features.
* This repository should be updated with build products on every merge to the code repo master branch using Travis.

This is how Psi4 builds their documentation automatically.

**Comment by langner** [comment]
yes yes yes every bit of automation here will be valuable, it seems there is never enough time to improve the docs, but it's an important user-facing part of the project

Please also see #5 where I was thinking about using autodoc for the generation.

**Comment by berquist** [comment]
Here is the strategy:

1. Move all the static material from cclib.github.io into cclib/docs.
2. Build docs for every PR.
3. On every merge to cclib:master, move built doc products into cclib.github.io:master (or other branch?) and commit.

Nothing autogenerated (even rst) should be committed into cclib/docs. This prevents polluting the code repo with HTML commits, and along with point 2 makes sure that the docs always build properly and never "on accident".

I thought we could copy [Psi4](https://github.com/psi4/psi4meta/tree/master/psinet-nightly) but have a separate docs server. We will make Travis do step 3.

**Comment by berquist** [comment]
I am in import hell with `coverage.py` because the directory and import structure has been changing. The long-term solution is probably to put it right inside `cclib/tests` so that each version of cclib has its own copy of `coverage.py`.

References to different versions or symlinks via different import mechanisms should be abstracted out of this file.

**Comment by berquist** [comment]
#36 gets the docs building again. Once that's in, a first PR will be simpler than proposed:
* Move all static/Sphinx stuff to `cclib`, deleting it from `cclib.github.io`
* Add `cclib.github.io` as a submodule inside `cclib/docs`
* Add a `deploy.sh` script that does the docs build, adds to `cclib.github.io`, and pushes

A later PR can do the Travis auto-build and testing.

**Comment by berquist** [comment]
I agree. Even simpler is to have a script/Travis clone the website repo, copy the built files into it, and then commit, but I was having authentication issues inside Travis. I'll try again.

**Comment by langner** [comment]
> I was having authentication issues inside Travis

What kind of issues?

**Comment by langner** [comment]
It's not entirely clear to me how the autogeneration will work in your mind. I thought maybe a github "project" would be good place to discuss this, but it seems those are just for task prioritization. Maybe you could recap your plans here?

**Comment by berquist** [comment]
Sure. I have been trying to keep it simple, because it's a lot to handle at once. You can think of each numbered step here as a separate PR.

1. Moving the docs and generators to the main repo needs to be done no matter what. So, rather than https://github.com/cclib/cclib.github.io/issues/5, everything will be in the same repo. The minimum work to do this is to move only the relevant files to the main repo, which is done in https://github.com/cclib/cclib/pull/556. No autogeneration done yet.
2. Delete everything in the website repo, commit, build the latest docs in the main repo, copy them over to the website repo, commit, and submit a PR. The website repo will now only ever contain a built website. This is done by hand at first so we know what it will look like, and we have a working copy of just the docs while the rest is set up.
3. Write a script (like `cclib/docs/deploy_to_gh_pages.sh`) that will clone the website repo, build the documentation, copy it over, commit, and push. Still no autogeneration.
4. Integrate calling the deployment script with Travis.

One reason to break this up is that 1-3 are easy, but I don't know how to do 4 yet. What I wrote earlier about when we should generate the docs is still true. Step 4 should:

* Build docs for every PR, so that if building them fails, the PR doesn't pass.
* On every successful merge to cclib:master, move built doc products into cclib.github.io:master and commit.

This is how I think we should break up the path to autogenerated documentation.

**Comment by langner** [comment]
Thanks for the overview, it's a good plan. I think we can handle 4 when we get there. I'm reasonably certain it can be done somehow. We can generate manually in the meantime.

My only comment is that we don't need the doc repo as a submodule in `cclib` which I think was still there in cclib/cclib#556.

**Comment by berquist** [comment]
That submodule is for the Sphinx theme, not the docs repo.

**Comment by berquist** [comment]
I figured it out how to do step 4; we can discuss it when it's time.

**Comment by langner** [comment]
> That submodule is for the Sphinx theme, not the docs repo.

Why do we need the sphinx theme in the main repo?

**Comment by langner** [comment]
> I figured it out how to do step 4; we can discuss it when it's time.

Yeah, let's break down the process into the steps you outlined and deal with them sequentially.

**Comment by berquist** [comment]
> Why do we need the sphinx theme in the main repo?

Once upon a time the Sphinx theme was modified from upstream (https://github.com/cclib/sphinx_rtd_theme/tree/cclib). I kept it because it was a submodule of the docs repo. If we're using the default rtd theme, then we can do away with it.

**Comment by langner** [comment]
> Why do we need the sphinx theme in the main repo?
> 
> Once upon a time the Sphinx theme was modified from upstream (https://github.com/cclib/sphinx_rtd_theme/tree/cclib). I kept it because it was a submodule of the docs repo. If we're using the default rtd theme, then we can do away with it.

Right. FTR, the modification had to do with supporting wide tables.

Can we keep a fork on the RTD repo instead? I would strongly prefer to not add unnecessary submodules to the library.

**Comment by berquist** [comment]
I can set it up so that we clone our fork at the correct branch/commit into the themes directory. This way, the fork is kept track of in a script rather than a submodule.

However, to have Travis manage this requires the same settings as auto-building the docs. Without going into much detail, the cclib org will make a personal access token that has `repo` permissions, and the hex token gets stored in a secret env variable on Travis. Otherwise Travis is incapable of pushing or even cloning other repos. _That_ is the authentication issue I was having.

In the meantime, I'll remove the submodule entirely and get things working with the "manual" script (step 3).

I will also investigate what the latest version of RTD looks like with cclib; it may be that our fork is obsolete.

**Comment by langner** [comment]
Thanks. I understand it's problematic, but I'm fairly certain it can be overcome. I'm not too worried about relaxing access to one repo since it'll be automatically generated content.

Have you look at things like https://github.com/Syntaf/travis-sphinx?

**Comment by berquist** [comment]
It's not problematic now that it's figured out. The script I wrote performs exactly the same workflow as that repo, but I'll take a further look.

**Comment by berquist** [comment]
Ok, I know how to automate everything using travis-sphinx. It does everything I was already doing, but it's good to use someone else's tested project. I also know how the code works so it can be patched if we don't like the behavior.

I've added a script that does 2/3, and the results are at https://github.com/cclib/cclib.github.io/tree/d07342fdba921d9ba355f32747d4c1c99d89a0f0. Switching over to serve docs from `gh-pages` will now work.

**Comment by langner** [comment]
Excellent!

**Comment by langner** [comment]
After #583 something seems to wrong with the doc build on merging. From [travis](https://travis-ci.org/cclib/cclib/jobs/423443075):

```
0.46s$ if [[ "${TRAVIS_BRANCH}" == master && "${TRAVIS_PULL_REQUEST}" == false ]];
then
    # Commits to master that are not pull requests, that is, only actual
    # addition of code to master, should deploy the documentation.
    bash ${TRAVIS_BUILD_DIR}/travis/deploy_docs_travis.bash
fi
Cloning into 'cclib.github.io'...
~/build/cclib/cclib/cclib.github.io ~/build/cclib/cclib
Copying built HTML...
cp: cannot stat ‘/home/travis/build/cclib/cclib/doc/sphinx/_build/html/*’: No such file or directory
The command "if [[ "${TRAVIS_BRANCH}" == master && "${TRAVIS_PULL_REQUEST}" == false ]];
then
    # Commits to master that are not pull requests, that is, only actual
    # addition of code to master, should deploy the documentation.
    bash ${TRAVIS_BUILD_DIR}/travis/deploy_docs_travis.bash
fi
" exited with 1.
cache.2
store build cache
0.00s
8.06schange detected (content changed, file is created, or file is deleted):
/home/travis/.cache/pip/http/0/a/0/f/a/0a0fa3965020c2e3afbbad5fbcae38ff11a524ab8327befdc1d582ee
/home/travis/.cache/pip/http/1/9/1/7/2/191721985457c3f06def70eaf404e7bb6dac69795150dc3e37f6bc33
/home/travis/.cache/pip/http/4/2/5/9/f/4259f1e2f5bcd9b3a38a52b0046ab46abbe04cd9d497d1617a790379
/home/travis/.cache/pip/http/4/4/e/5/b/44e5b11a6caa92636d8ccfe658d420ba4ed8f67f7f4e835b214255aa
/home/travis/.cache/pip/http/4/5/9/7/6/45976c194ebb9180bf7410d422bcad653ce13bcb64dda0d07d72b09d
/home/travis/.cache/pip/http/4/b/e/0/7/4be07d3ac353e38d2c9e3a257cad36ee2a758fc88b4cd5f0d479a5a9
/home/travis/.cache/pip/http/5/9/3/8/0/59380b71dcd1fa0e5d423801f68589f3beee17a7c22eb1e2b621bbac
/home/travis/.cache/pip/http/5/b/e/1/1/5be1135dcb7a68de4427c71119ff80f66b671ae16db019966320bd6b
/home/travis/.cache/pip/http/6/a/c/f/e/6acfe714f42ee2380289743b5c1f9e2bea4a2a556b80cb91e6a409d2
/home/travis/.cache/pip/http/6/b/7/0/1/6b701ea6dd0ab0b2838453e7348c98d3060441eb36fea66fc36a5bee
/home/travis/.cache/pip/http/6/c/c/a/3/6c
...
changes detected, packing new archive
uploading archive
Done. Your build exited with 1.
```

**Comment by langner** [comment]
Potential fix in cclib/cclib#609

**Comment by langner** [comment]
Works!

**Comment by berquist** [comment]
For reference, this is the first automated commit: https://github.com/cclib/cclib.github.io/commit/6f270b29a450daccc395fd46d2e8b1a339665ca7
